File: apps/desktop/src-tauri/src/git_integration.rs
use serde::Serialize;
use std::{fs, path::Path};
#[path = "git_integration/config.rs"]
mod config;
#[path = "git_integration/support.rs"]
mod support;
#[cfg(test)]
use config::stable_cli_path;
use config::{
    canonical_repo_root, config_value, driver_commands, driver_configured, ensure_repo,
    find_in_path, is_executable, write_gitattributes_line, write_global_config, write_repo_config,
    WriteConfigReport,
};
use support::{log_event, new_request_id, run_git, stderr_or_stdout};
const BARE_DIFF_TEXTCONV: &str = "aimd git-diff";
const BARE_MERGE_DRIVER: &str = "aimd git-merge %O %A %B %P";
#[cfg(not(windows))]
const STABLE_DIFF_TEXTCONV: &str = "/usr/local/bin/aimd git-diff";
#[cfg(not(windows))]
const STABLE_MERGE_DRIVER: &str = "/usr/local/bin/aimd git-merge %O %A %B %P";
const MERGE_NAME: &str = "AIMD merge driver";
const GITATTRIBUTES_LINE: &str = "*.aimd diff=aimd merge=aimd";
#[cfg(not(windows))]
const STABLE_CLI_PATH: &str = "/usr/local/bin/aimd";
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitIntegrationStatus {
    pub request_id: String,
    pub git_installed: bool,
    pub git_error: Option<String>,
    pub cli_in_path: bool,
    pub cli_path: Option<String>,
    pub stable_cli_path: String,
    pub stable_cli_exists: bool,
    pub stable_cli_executable: bool,
    pub stable_cli_error: Option<String>,
    pub repo_path: Option<String>,
    pub repo_is_git: bool,
    pub repo_error: Option<String>,
    pub repo_path_requested: bool,
    pub gitattributes_present: bool,
    pub gitattributes_configured: bool,
    pub repo_driver_configured: bool,
    pub global_driver_configured: bool,
    pub driver_command_source: String,
    pub expected_textconv: String,
    pub expected_merge_driver: String,
    pub global_textconv: Option<String>,
    pub global_cache_textconv: Option<String>,
    pub global_merge_name: Option<String>,
    pub global_merge_driver: Option<String>,
    pub repo_textconv: Option<String>,
    pub repo_cache_textconv: Option<String>,
    pub repo_merge_name: Option<String>,
    pub repo_merge_driver: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitDoctorResult {
    pub request_id: String,
    pub ok: bool,
    pub messages: Vec<String>,
    pub suggestions: Vec<String>,
    pub status: GitIntegrationStatus,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitIntegrationActionResult {
    pub request_id: String,
    pub ok: bool,
    pub title: String,
    pub message: String,
    pub details: Vec<String>,
    pub status: GitIntegrationStatus,
}

#[tauri::command]
pub fn git_integration_status(repo_path: Option<String>) -> Result<GitIntegrationStatus, String> {
    let request_id = new_request_id("status");
    let status = status_impl(repo_path.as_deref(), request_id.clone());
    log_event(
        &request_id,
        "status",
        "status",
        repo_path.as_deref(),
        None,
        None,
        None,
        "ok",
        "status checked",
    );
    Ok(status)
}

#[tauri::command]
pub fn git_integration_enable_global() -> Result<GitIntegrationActionResult, String> {
    let request_id = new_request_id("global-enable");
    let result = write_global_config(true, &request_id);
    action_result(
        request_id,
        "global",
        None,
        result,
        "全局 Git driver 已启用",
        "全局 Git driver 启用失败",
        None,
    )
}

#[tauri::command]
pub fn git_integration_disable_global() -> Result<GitIntegrationActionResult, String> {
    let request_id = new_request_id("global-disable");
    let result = write_global_config(false, &request_id);
    action_result(
        request_id,
        "global",
        None,
        result,
        "全局 Git driver 已禁用",
        "全局 Git driver 禁用失败",
        None,
    )
}

#[tauri::command]
pub fn git_integration_enable_repo(
    repo_path: String,
) -> Result<GitIntegrationActionResult, String> {
    let request_id = new_request_id("repo-enable");
    let repo = ensure_repo(&repo_path)?;
    let result = write_repo_config(&repo, true, &request_id);
    action_result(
        request_id,
        "repo",
        Some(repo_path.as_str()),
        result,
        "当前仓库 Git driver 已启用",
        "当前仓库 Git driver 启用失败",
        Some(repo_path.as_str()),
    )
}

#[tauri::command]
pub fn git_integration_disable_repo(
    repo_path: String,
) -> Result<GitIntegrationActionResult, String> {
    let request_id = new_request_id("repo-disable");
    let repo = ensure_repo(&repo_path)?;
    let result = write_repo_config(&repo, false, &request_id);
    action_result(
        request_id,
        "repo",
        Some(repo_path.as_str()),
        result,
        "当前仓库 Git driver 已禁用",
        "当前仓库 Git driver 禁用失败",
        Some(repo_path.as_str()),
    )
}

#[tauri::command]
pub fn git_integration_write_gitattributes(
    repo_path: String,
) -> Result<GitIntegrationActionResult, String> {
    let request_id = new_request_id("gitattributes");
    let repo = ensure_repo(&repo_path)?;
    let result = write_gitattributes_line(&repo).map(|details| WriteConfigReport { details });
    action_result(
        request_id,
        "gitattributes",
        Some(repo_path.as_str()),
        result,
        ".gitattributes 已写入",
        ".gitattributes 写入失败",
        Some(repo_path.as_str()),
    )
}

#[tauri::command]
pub fn git_integration_doctor(repo_path: Option<String>) -> Result<GitDoctorResult, String> {
    let request_id = new_request_id("doctor");
    let status = status_impl(repo_path.as_deref(), request_id.clone());
    let mut messages = Vec::new();
    let mut suggestions = Vec::new();
    if !status.git_installed {
        messages.push(
            status
                .git_error
                .clone()
                .unwrap_or_else(|| "Git 不可用".to_string()),
        );
        suggestions.push("安装或修复 Git 后重新检查".to_string());
    }
    if status.stable_cli_executable {
        if !status.cli_in_path {
            messages.push(format!(
                "aimd 不在 App PATH 中，已使用 {} 作为稳定入口",
                status.stable_cli_path
            ));
        }
    } else {
        messages.push(
            status
                .stable_cli_error
                .clone()
                .unwrap_or_else(|| format!("{} 不可执行", status.stable_cli_path)),
        );
        suggestions.push(format!(
            "安装 AIMD Windows 安装包，确保 {} 可执行",
            status.stable_cli_path
        ));
    }
    if !status.repo_path_requested {
        messages.push("未设置仓库路径，仓库 driver 和 .gitattributes 检查不适用".to_string());
        suggestions.push("填写仓库路径后可启用当前仓库 Git 集成".to_string());
    } else if !status.repo_is_git {
        messages.push(
            status
                .repo_error
                .clone()
                .unwrap_or_else(|| "当前路径不是 Git 仓库".to_string()),
        );
        suggestions.push("填写有效 Git 仓库路径".to_string());
    }
    if status.repo_is_git && !status.gitattributes_configured {
        messages.push(".gitattributes 未包含 *.aimd diff=aimd merge=aimd".to_string());
        suggestions.push("点击写入 .gitattributes".to_string());
    }
    if !status.global_driver_configured && !status.repo_driver_configured {
        messages.push("尚未配置 AIMD Git driver".to_string());
        suggestions.push("点击启用全局 Git 集成或启用当前仓库 Git 集成".to_string());
    }
    log_event(
        &request_id,
        "doctor",
        "doctor",
        repo_path.as_deref(),
        None,
        None,
        None,
        "ok",
        "doctor checked",
    );
    Ok(GitDoctorResult {
        request_id,
        ok: messages.is_empty(),
        messages,
        suggestions,
        status,
    })
}

fn status_impl(repo_path: Option<&str>, request_id: String) -> GitIntegrationStatus {
    let git_version = run_git(None, &["--version"]);
    let git_installed = git_version
        .as_ref()
        .map(|o| o.status.success())
        .unwrap_or(false);
    let git_error = match git_version {
        Ok(out) if out.status.success() => None,
        Ok(out) => Some(stderr_or_stdout(&out)),
        Err(err) => Some(err),
    };
    let cli_path_buf = find_in_path("aimd");
    let cli_path = cli_path_buf
        .as_ref()
        .map(|p| p.to_string_lossy().to_string());
    let stable = config::stable_cli_path();
    let stable_cli_exists = stable.exists();
    let stable_cli_executable = is_executable(&stable);
    let commands = driver_commands(
        cli_path_buf.as_deref(),
        stable_cli_executable.then_some(stable.as_path()),
    );
    let stable_cli_error = if stable_cli_exists && !stable_cli_executable {
        Some(format!("{} 存在但不可执行", stable.display()))
    } else if !stable_cli_exists {
        Some(format!("{} 不存在", stable.display()))
    } else {
        None
    };

    let repo_path_requested = repo_path.map(|p| !p.trim().is_empty()).unwrap_or(false);
    let repo = repo_path
        .filter(|p| !p.trim().is_empty())
        .and_then(|p| canonical_repo_root(Path::new(p)).ok());
    let repo_error = if repo_path.is_some() && repo.is_none() {
        Some("不是 Git 仓库或无法访问 .git".to_string())
    } else {
        None
    };
    let gitattributes = repo.as_ref().map(|r| r.join(".gitattributes"));
    let gitattributes_present = gitattributes.as_ref().map(|p| p.exists()).unwrap_or(false);
    let gitattributes_configured = gitattributes
        .as_ref()
        .and_then(|p| fs::read_to_string(p).ok())
        .map(|s| s.lines().any(|line| line.trim() == GITATTRIBUTES_LINE))
        .unwrap_or(false);

    GitIntegrationStatus {
        request_id,
        git_installed,
        git_error,
        cli_in_path: cli_path.is_some(),
        cli_path,
        stable_cli_path: stable.to_string_lossy().to_string(),
        stable_cli_exists,
        stable_cli_executable,
        stable_cli_error,
        repo_path: repo.as_ref().map(|p| p.to_string_lossy().to_string()),
        repo_is_git: repo.is_some(),
        repo_error,
        repo_path_requested,
        gitattributes_present,
        gitattributes_configured,
        repo_driver_configured: repo
            .as_ref()
            .map(|r| driver_configured(Some(r), &commands))
            .unwrap_or(false),
        global_driver_configured: driver_configured(None, &commands),
        driver_command_source: commands.source,
        expected_textconv: commands.textconv,
        expected_merge_driver: commands.merge_driver,
        global_textconv: config_value(None, "diff.aimd.textconv"),
        global_cache_textconv: config_value(None, "diff.aimd.cachetextconv"),
        global_merge_name: config_value(None, "merge.aimd.name"),
        global_merge_driver: config_value(None, "merge.aimd.driver"),
        repo_textconv: repo
            .as_ref()
            .and_then(|r| config_value(Some(r), "diff.aimd.textconv")),
        repo_cache_textconv: repo
            .as_ref()
            .and_then(|r| config_value(Some(r), "diff.aimd.cachetextconv")),
        repo_merge_name: repo
            .as_ref()
            .and_then(|r| config_value(Some(r), "merge.aimd.name")),
        repo_merge_driver: repo
            .as_ref()
            .and_then(|r| config_value(Some(r), "merge.aimd.driver")),
    }
}

fn action_result(
    request_id: String,
    scope: &str,
    repo_path: Option<&str>,
    result: Result<WriteConfigReport, String>,
    success_title: &str,
    failure_title: &str,
    status_repo_path: Option<&str>,
) -> Result<GitIntegrationActionResult, String> {
    match result {
        Ok(report) => {
            let status = status_impl(status_repo_path, request_id.clone());
            log_event(
                &request_id,
                "action",
                scope,
                repo_path,
                None,
                None,
                None,
                "ok",
                success_title,
            );
            Ok(GitIntegrationActionResult {
                request_id,
                ok: true,
                title: success_title.to_string(),
                message: "操作已完成并通过验证".to_string(),
                details: report.details,
                status,
            })
        }
        Err(err) => {
            log_event(
                &request_id,
                "action",
                scope,
                repo_path,
                None,
                None,
                None,
                "failed",
                &err,
            );
            Err(format!("{failure_title}: {err}（requestId: {request_id}）"))
        }
    }
}

#[cfg(test)]
#[path = "git_integration/tests.rs"]
mod tests;

File: apps/desktop/src/settings/git-integration.ts
import { invoke } from "@tauri-apps/api/core";

type GitIntegrationStatus = {
  requestId?: string;
  gitInstalled: boolean;
  gitError?: string;
  cliInPath: boolean;
  cliPath?: string;
  stableCliPath?: string;
  stableCliExists: boolean;
  stableCliExecutable: boolean;
  stableCliError?: string;
  repoPath?: string;
  repoIsGit: boolean;
  repoError?: string;
  repoPathRequested?: boolean;
  gitattributesPresent: boolean;
  gitattributesConfigured: boolean;
  repoDriverConfigured: boolean;
  globalDriverConfigured: boolean;
  driverCommandSource?: string;
  expectedTextconv: string;
  expectedMergeDriver: string;
  globalTextconv?: string;
  globalCacheTextconv?: string;
  globalMergeName?: string;
  globalMergeDriver?: string;
  repoTextconv?: string;
  repoCacheTextconv?: string;
  repoMergeName?: string;
  repoMergeDriver?: string;
};

type GitDoctorResult = {
  requestId?: string;
  ok: boolean;
  messages: string[];
  suggestions?: string[];
  status: GitIntegrationStatus;
};

type GitIntegrationActionResult = {
  requestId?: string;
  ok: boolean;
  title: string;
  message: string;
  details?: string[];
  status: GitIntegrationStatus;
};

type LineState = "ok" | "missing" | "notConfigured" | "notApplicable" | "error" | "checking";

export function gitIntegrationSectionHTML() {
  return `
    <section id="settings-panel-git" class="settings-section" data-section="git" role="tabpanel" aria-labelledby="settings-tab-git" hidden>
      <header class="settings-section-head">
        <h2>Git 集成</h2>
        <p>安装、检测、修复和卸载 AIMD 的 .aimd diff / merge driver。</p>
      </header>

      <label class="field">
        <span class="field-label">仓库路径 <em class="field-optional">可选</em></span>
        <input id="git-repo-path" type="text" autocomplete="off" spellcheck="false" placeholder="/path/to/git/repo" />
        <span class="field-hint">留空时只检查 Git、CLI 和全局 driver；仓库操作需要填写 Git 仓库路径。</span>
      </label>

      <div id="git-integration-status" class="git-integration-status" aria-live="polite"></div>

      <div class="git-integration-actions">
        <button id="git-refresh" class="secondary-btn sm" type="button">检查/修复</button>
        <button id="git-enable-repo" class="secondary-btn sm" type="button">启用当前仓库 Git 集成</button>
        <button id="git-enable-global" class="secondary-btn sm" type="button">启用全局 Git 集成</button>
        <button id="git-write-attrs" class="secondary-btn sm" type="button">写入 .gitattributes</button>
        <button id="git-disable-repo" class="secondary-btn sm" type="button">禁用当前仓库 Git 集成</button>
        <button id="git-disable-global" class="secondary-btn sm" type="button">禁用全局 Git 集成</button>
      </div>
    </section>
  `;
}

export function setupGitIntegration(root: HTMLElement) {
  const $ = <T extends HTMLElement>(selector: string) => root.querySelector<T>(selector)!;
  const repoPathEl = $<HTMLInputElement>("#git-repo-path");
  const statusEl = $<HTMLElement>("#git-integration-status");
  const refreshBtn = $<HTMLButtonElement>("#git-refresh");
  const enableRepoBtn = $<HTMLButtonElement>("#git-enable-repo");
  const enableGlobalBtn = $<HTMLButtonElement>("#git-enable-global");
  const writeAttrsBtn = $<HTMLButtonElement>("#git-write-attrs");
  const disableRepoBtn = $<HTMLButtonElement>("#git-disable-repo");
  const disableGlobalBtn = $<HTMLButtonElement>("#git-disable-global");
  let busy = false;

  function repoPathValue(): string | null {
    const value = repoPathEl.value.trim();
    return value ? value : null;
  }

  function statusLine(label: string, state: LineState, detail = "") {
    const text: Record<LineState, string> = {
      ok: "OK",
      missing: "缺失",
      notConfigured: "未配置",
      notApplicable: "不适用",
      error: "失败",
      checking: "检查中",
    };
    const ok = state === "ok" || state === "notApplicable";
    return `<div class="git-status-line" data-ok="${String(ok)}" data-state="${state}"><strong>${label}</strong><span>${text[state]}${detail ? ` · ${escapeHTML(detail)}` : ""}</span></div>`;
  }

  function valueLine(label: string, value?: string) {
    return `<div class="git-status-line" data-ok="${String(Boolean(value))}" data-state="${value ? "ok" : "notConfigured"}"><strong>${label}</strong><span>${value ? escapeHTML(value) : "未配置"}</span></div>`;
  }

  function renderAction(result: GitIntegrationActionResult) {
    renderStatus(result.status, [], result);
  }

  function renderStatus(status: GitIntegrationStatus, messages: string[] = [], action?: GitIntegrationActionResult, suggestions: string[] = []) {
    const repoRequested = Boolean(status.repoPathRequested ?? status.repoPath);
    const systemCliOk = status.stableCliExists && status.stableCliExecutable;
    const repoState: LineState = repoRequested ? (status.repoIsGit ? "ok" : "error") : "notConfigured";
    const attrsState: LineState = !repoRequested ? "notApplicable" : status.gitattributesConfigured ? "ok" : "notConfigured";
    const repoDriverState: LineState = !repoRequested ? "notApplicable" : status.repoDriverConfigured ? "ok" : "notConfigured";
    const cliDetail = status.cliInPath && status.cliPath
      ? `${status.cliPath}（来自 PATH）`
      : systemCliOk
        ? `${status.stableCliPath || status.expectedTextconv.replace(/\s+git-diff$/, "")}（稳定安装入口，App 会使用它）`
        : status.stableCliError || "需要安装 AIMD PKG，或在开发环境 PATH 中提供 aimd";
    const globalDetail = status.globalDriverConfigured
      ? "已启用，.aimd diff / merge 会调用 AIMD"
      : "未启用；点击“启用全局 Git 集成”";
    const repoDetail = !repoRequested
      ? "未填写仓库路径；只影响当前仓库的操作会先禁用"
      : status.repoIsGit
        ? status.repoPath || "已识别 Git 仓库"
        : status.repoError || "不是 Git 仓库";
    const repoDriverDetail = !repoRequested
      ? "填写仓库路径后可检查/启用"
      : status.repoDriverConfigured
        ? "已启用，仅作用于这个仓库"
        : "未启用；可点击“启用当前仓库 Git 集成”";
    const attrsDetail = !repoRequested
      ? "填写仓库路径后可写入"
      : status.gitattributesConfigured
        ? "已配置 *.aimd"
        : status.gitattributesPresent
          ? "文件存在，但还没有配置 *.aimd"
          : "未写入；可点击“写入 .gitattributes”";
    const lines = [
      action ? `<div class="git-status-line" data-ok="${String(action.ok)}"><strong>${escapeHTML(action.title)}</strong><span>${escapeHTML(action.message)}${action.requestId ? ` · ${escapeHTML(action.requestId)}` : ""}</span></div>` : "",
      statusLine("Git", status.gitInstalled ? "ok" : "error", status.gitError || ""),
      statusLine("AIMD CLI", status.cliInPath || systemCliOk ? "ok" : "missing", cliDetail),
      statusLine("全局 Git 集成", status.globalDriverConfigured ? "ok" : "notConfigured", globalDetail),
      statusLine("当前仓库", repoState, repoDetail),
      statusLine("当前仓库 Git 集成", repoDriverState, repoDriverDetail),
      statusLine(".gitattributes", attrsState, attrsDetail),
      statusLine("Git driver 命令", "ok", `${status.driverCommandSource === "stable" ? "系统 CLI" : "PATH"} · ${status.expectedTextconv}`),
    ];
    if (status.globalTextconv || status.globalMergeDriver || status.repoTextconv || status.repoMergeDriver) {
      lines.push(
        valueLine("全局 diff 命令", status.globalTextconv),
        valueLine("全局 merge 命令", status.globalMergeDriver),
        repoRequested ? valueLine("仓库 diff 命令", status.repoTextconv) : "",
        repoRequested ? valueLine("仓库 merge 命令", status.repoMergeDriver) : "",
      );
    }
    if (action?.details?.length) {
      lines.push(`<div class="git-status-line" data-ok="true"><strong>操作详情</strong><span>${action.details.map(escapeHTML).join("；")}</span></div>`);
    }
    if (messages.length > 0) {
      lines.push(`<div class="git-status-line" data-ok="false"><strong>诊断</strong><span>${messages.map(escapeHTML).join("；")}</span></div>`);
    }
    if (suggestions.length > 0) {
      lines.push(`<div class="git-status-line" data-ok="true"><strong>建议</strong><span>${suggestions.map(escapeHTML).join("；")}</span></div>`);
    }
    statusEl.innerHTML = lines.filter(Boolean).join("");
    enableRepoBtn.disabled = busy || !repoRequested || !status.repoIsGit;
    disableRepoBtn.disabled = busy || !repoRequested || !status.repoIsGit;
    writeAttrsBtn.disabled = busy || !repoRequested || !status.repoIsGit;
  }

  function renderError(err: unknown) {
    statusEl.innerHTML = `<div class="git-status-line" data-ok="false"><strong>Git 集成</strong><span>${escapeHTML(String(err))}</span></div>`;
  }

  function syncButtons() {
    [refreshBtn, enableGlobalBtn, disableGlobalBtn].forEach((btn) => { btn.disabled = busy; });
  }

  async function refresh() {
    if (busy) return;
    busy = true;
    syncButtons();
    statusEl.textContent = "检查中...";
    try {
      const status = await invoke<GitIntegrationStatus>("git_integration_status", { repoPath: repoPathValue() });
      busy = false;
      renderStatus(status);
    } catch (err) {
      renderError(err);
    } finally {
      busy = false;
      syncButtons();
    }
  }

  async function doctor() {
    if (busy) return;
    busy = true;
    syncButtons();
    statusEl.textContent = "检查中...";
    try {
      const result = await invoke<GitDoctorResult>("git_integration_doctor", { repoPath: repoPathValue() });
      busy = false;
      renderStatus(result.status, result.messages, undefined, result.suggestions || []);
    } catch (err) {
      renderError(err);
    } finally {
      busy = false;
      syncButtons();
    }
  }

  async function confirmGitConfigChange(): Promise<boolean> {
    const message = "启用后，Git 在 diff/merge .aimd 文件时会调用 AIMD 命令。\n这会修改 Git 配置，可随时在这里关闭。";
    try {
      return await invoke<boolean>("confirm_git_config_change", { message });
    } catch {
      // 仅用于 vite-only 开发态 / e2e 兜底；Tauri 实跑必须走 Rust 原生对话框。
      return window.confirm(message);
    }
  }

  async function runAction(command: string, args: Record<string, unknown> = {}, confirmWrite = false) {
    if (busy) return;
    if (confirmWrite && !(await confirmGitConfigChange())) {
      statusEl.innerHTML = `<div class="git-status-line" data-ok="true"><strong>已取消</strong><span>未修改 Git 配置</span></div>`;
      return;
    }
    busy = true;
    syncButtons();
    statusEl.textContent = "处理中...";
    try {
      const result = await invoke<GitIntegrationActionResult>(command, args);
      busy = false;
      renderAction(result);
    } catch (err) {
      renderError(err);
    } finally {
      busy = false;
      syncButtons();
    }
  }

  function requireRepoPath(): string | null {
    const repoPath = repoPathValue();
    if (!repoPath) renderError("请先填写仓库路径");
    return repoPath;
  }

  repoPathEl.addEventListener("change", () => { void refresh(); });
  refreshBtn.addEventListener("click", () => { void doctor(); });
  enableGlobalBtn.addEventListener("click", () => { void runAction("git_integration_enable_global", {}, true); });
  disableGlobalBtn.addEventListener("click", () => { void runAction("git_integration_disable_global"); });
  enableRepoBtn.addEventListener("click", () => {
    const repoPath = requireRepoPath();
    if (repoPath) void runAction("git_integration_enable_repo", { repoPath }, true);
  });
  disableRepoBtn.addEventListener("click", () => {
    const repoPath = requireRepoPath();
    if (repoPath) void runAction("git_integration_disable_repo", { repoPath });
  });
  writeAttrsBtn.addEventListener("click", () => {
    const repoPath = requireRepoPath();
    if (repoPath) void runAction("git_integration_write_gitattributes", { repoPath });
  });

  return { refresh, syncButtons };
}

function escapeHTML(value: string) {
  return value.replace(/[&<>"']/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c]!));
}
